### PR TITLE
AES-CTR with re-seeding (#21740)

### PR DESCRIPTION
**Pull request by gosubpl**
Introduces an AES-CTR re-seeding PRNG to replace the one from maths.uncommon.

### CODE REVIEW

**Comment by gosubpl** [comment]
Refs #21740

File: akka-remote/src/main/scala/akka/remote/security/provider/AESCounterBuiltinCTRRNG.scala
Comment (patriknw) [inline akka-remote/src/main/scala/akka/remote/security/provider/AESCounterBuiltinCTRRNG.scala]: this condition was already checked above

File: akka-remote/src/main/scala/akka/remote/security/provider/AESCounterBuiltinCTRRNG.scala
Comment (patriknw) [inline akka-remote/src/main/scala/akka/remote/security/provider/AESCounterBuiltinCTRRNG.scala]: I don't fully understand the intention here. current thread is anyway blocked by the below Await so I don't see the purpose of delegate the `generateSeed` to another thread. Other threads are blocked on the lock until this is completed.

File: akka-remote/src/main/scala/akka/remote/security/provider/AES256CounterBuiltinRNG.scala
Comment (gosubpl) [inline akka-remote/src/main/scala/akka/remote/security/provider/AES256CounterBuiltinRNG.scala]: fixed

File: akka-remote/src/main/scala/akka/remote/security/provider/AES256CounterBuiltinRNG.scala
Comment (gosubpl) [inline akka-remote/src/main/scala/akka/remote/security/provider/AES256CounterBuiltinRNG.scala]: this was a direct copy of uncommons.math file, used for performance comparisons, is removed now, as it is not needed in akka

File: akka-remote/src/main/scala/akka/remote/security/provider/AESCounterBuiltinCTRRNG.scala
Comment (gosubpl) [inline akka-remote/src/main/scala/akka/remote/security/provider/AESCounterBuiltinCTRRNG.scala]: fixed

File: akka-remote/src/main/scala/akka/remote/security/provider/AESCounterBuiltinCTRRNG.scala
Comment (gosubpl) [inline akka-remote/src/main/scala/akka/remote/security/provider/AESCounterBuiltinCTRRNG.scala]: fixed

File: akka-remote/src/main/scala/akka/remote/security/provider/AESCounterBuiltinCTRRNG.scala
Comment (gosubpl) [inline akka-remote/src/main/scala/akka/remote/security/provider/AESCounterBuiltinCTRRNG.scala]: not exactly, we have checked in line 81 if the `bitsSinceSeeding` exceeded `RESEEDING_THRESHOLD` which should trigger requesting for new entropy. As new entropy request would block, we run this on separate thread (ekhm, Future) thus avoiding blocking in the main thread (agreed, this should be on io dispatcher, not the default one).
However, if (what is very unlikely) we don't get the entropy back and we have reached another, higher limit of `RESEEDING_DEADLINE`, then we need to block (hence `Await`) and wait for new entropy to arrive, as proceeding would be unsafe

File: akka-remote/src/main/scala/akka/remote/security/provider/AESCounterBuiltinCTRRNG.scala
Comment (gosubpl) [inline akka-remote/src/main/scala/akka/remote/security/provider/AESCounterBuiltinCTRRNG.scala]: fixed

File: akka-remote/src/main/scala/akka/remote/security/provider/AES256CounterBuiltinRNG.scala
Comment (gosubpl) [inline akka-remote/src/main/scala/akka/remote/security/provider/AES256CounterBuiltinRNG.scala]: however, please note that I have renamed the new proposed algorithm from `AESCounterBuiltinCTRRNG` to `AESCounterBuiltinRNG` (which is a good name in itself), so the file itself will stay, albeit with different content

File: akka-remote/src/main/scala/akka/remote/security/provider/AESCounterBuiltinCTRRNG.scala
Comment (gosubpl) [inline akka-remote/src/main/scala/akka/remote/security/provider/AESCounterBuiltinCTRRNG.scala]: working on this

File: akka-remote/src/main/scala/akka/remote/security/provider/AESCounterBuiltinCTRRNG.scala
Comment (gosubpl) [inline akka-remote/src/main/scala/akka/remote/security/provider/AESCounterBuiltinCTRRNG.scala]: working on this - I need to run benchmarks to see what are the trade-offs

File: akka-remote/src/main/scala/akka/remote/security/provider/AESCounterBuiltinCTRRNG.scala
Comment (gosubpl) [inline akka-remote/src/main/scala/akka/remote/security/provider/AESCounterBuiltinCTRRNG.scala]: Hmmm... will add some clarifying comments - generally the flow looks like this:
1) if the `RESEEDING_THRESHOLD` is exceeded (new entropy is needed), create a Future running the new seed generator
2) cyclically check whether this future hasn't completed (in a non-blocking way). If it has - retrieve the value.
3) check whether by any chance `RESEEDING_DEADLINE` was not exceeded - if it has, we _must_ re-seed - in that case, activate `Await` which would block

File: akka-remote/src/main/scala/akka/remote/security/provider/AESCounterBuiltinCTRRNG.scala
Comment (gosubpl) [inline akka-remote/src/main/scala/akka/remote/security/provider/AESCounterBuiltinCTRRNG.scala]: The thing inside the future is basically just a read from `/dev/random`, so with high probability should succeed, and if it doesn't, then what guarantee we have that subsequent call that will attempt the same read will succeed? In any way, at this stage we cannot proceed without the additional entropy. But for added security we could add some re-try logic in case this fails.

Comment (gosubpl) [inline akka-remote/src/main/scala/akka/remote/security/provider/AESCounterBuiltinCTRRNG.scala]: The thing inside the future is basically just a read from `/dev/random` (or `/dev/urandom` on BSD or system call on Windows). If it does not return in reasonable timeframe, then probably something is very wrong. But I could imagine a retry logic backed up by an exception after a couple of failed attempts, WDYT?

File: akka-remote/src/main/scala/akka/remote/security/provider/AESCounterBuiltinCTRRNG.scala
Comment (gosubpl) [inline akka-remote/src/main/scala/akka/remote/security/provider/AESCounterBuiltinCTRRNG.scala]: yes, agreed; will refactor if synchronized is not a solution

File: akka-remote/src/main/scala/akka/remote/security/provider/AESCounterBuiltinRNG.scala
Comment (gosubpl) [inline akka-remote/src/main/scala/akka/remote/security/provider/AESCounterBuiltinRNG.scala]: working on introducing separate execution context

File: akka-remote/src/main/scala/akka/remote/security/provider/AESCounterBuiltinRNG.scala
Comment (gosubpl) [inline akka-remote/src/main/scala/akka/remote/security/provider/AESCounterBuiltinRNG.scala]: analyzing whether not replace with `synchronized` block

File: akka-remote/src/main/scala/akka/remote/security/provider/AESCounterBuiltinRNG.scala
Comment (gosubpl) [inline akka-remote/src/main/scala/akka/remote/security/provider/AESCounterBuiltinRNG.scala]: not sure if this is a good name

File: akka-remote/src/main/scala/akka/remote/security/provider/AESCounterBuiltinRNG.scala
Comment (gosubpl) [inline akka-remote/src/main/scala/akka/remote/security/provider/AESCounterBuiltinRNG.scala]: maybe requires some logic for re-try and final fail with exception, but basically what's inside the Future is just a read from `/dev/random` so with high probability will not fail, WDYT?

File: akka-remote/src/main/scala/akka/remote/security/provider/AESCounterBuiltinRNG.scala
Comment (gosubpl) [inline akka-remote/src/main/scala/akka/remote/security/provider/AESCounterBuiltinRNG.scala]: if lock is kept, another layer of try-finally will be added

**Comment by gosubpl** [comment]
@patriknw many thanks for the insightful comments. I am still working on improving this code (especially on dispatcher and lock), will ping you when it's ready for another review. In the meantime would be great if @drewhk could chime in on the re-seeding logic, especially on what could be a good value for `RESEEDING_THRESHOLD` (which exceeding initiates re-seeding in a non-blocking way as opposed to exceeding `RESEEDING_DEADLINE` that results in blocking)

File: akka-remote/src/main/scala/akka/remote/security/provider/AES256CounterBuiltinRNG.scala
Comment (gosubpl) [inline akka-remote/src/main/scala/akka/remote/security/provider/AES256CounterBuiltinRNG.scala]: please forgive the confusion

File: akka-remote/src/main/scala/akka/remote/security/provider/AESCounterBuiltinRNG.scala
Comment (patriknw) [inline akka-remote/src/main/scala/akka/remote/security/provider/AESCounterBuiltinRNG.scala]: I'm always scared about blocking forever on something because there is no sign of why the application is halting. Yeah, it should always return within reasonable time, but I'd rather have a long timeout (e.g. 5 minutes) with logging in case of failure than trusting an external "service".

Retrying might be overkill.

File: akka-remote/src/main/scala/akka/remote/security/provider/AESCounterBuiltinRNG.scala
Comment (patriknw) [inline akka-remote/src/main/scala/akka/remote/security/provider/AESCounterBuiltinRNG.scala]: ah, in my previous review I misread and though it was the same constant

File: akka-remote/src/main/scala/akka/remote/security/provider/AES128CounterSecureRNG.scala
Comment (gosubpl) [inline akka-remote/src/main/scala/akka/remote/security/provider/AES128CounterSecureRNG.scala]: `ThreadPoolExecutor` with 1 thread. This is enough as if the future does not complete, the generator will block and will not submit another future.

File: akka-remote/src/main/scala/akka/remote/security/provider/AES128CounterSecureRNG.scala
Comment (gosubpl) [inline akka-remote/src/main/scala/akka/remote/security/provider/AES128CounterSecureRNG.scala]: This method must be implemented to comply with `SecureRandomSpi` API

File: akka-remote/src/main/scala/akka/remote/security/provider/AESCounterBuiltinRNG.scala
Comment (gosubpl) [inline akka-remote/src/main/scala/akka/remote/security/provider/AESCounterBuiltinRNG.scala]: Synchronized is indeed faster (benchmarked) for uncontended access (which is the case here). Plus the control flow is easier (no need for unlocking in finally)

File: akka-remote/src/main/scala/akka/remote/security/provider/AESCounterBuiltinRNG.scala
Comment (gosubpl) [inline akka-remote/src/main/scala/akka/remote/security/provider/AESCounterBuiltinRNG.scala]: This is the only option of logging - we don't have access to `ActorSystem` here

File: akka-remote/src/main/scala/akka/remote/security/provider/AESCounterBuiltinRNG.scala
Comment (gosubpl) [inline akka-remote/src/main/scala/akka/remote/security/provider/AESCounterBuiltinRNG.scala]: Default time-out is 5 minutes. Something would need to be seriously wrong to exceed

File: akka-remote/src/test/scala/akka/remote/serialization/RandomReplacementSpec.scala
Comment (gosubpl) [inline akka-remote/src/test/scala/akka/remote/serialization/RandomReplacementSpec.scala]: A spec to test whether the new generator generates the same sequence as the original one. This spec can be removed after review.

File: akka-bench-jmh/src/main/scala/akka/remote/RandomReplacementBenchmark.scala
Comment (gosubpl) [inline akka-bench-jmh/src/main/scala/akka/remote/RandomReplacementBenchmark.scala]: Benchmark for reference purposes. Can be removed before merge.

**Comment by gosubpl** [comment]
@patriknw @drewhk this PR is now ready for review and almost ready for merge. The things pending are:
* is `RESEEDING_THRESHOLD = 1000000000L` (one billion) ok? this is the threshold at which re-seeding will be started. Shouldn't be too low to avoid too frequent re-seeds. The safe limit (defined by `RESEEDING_DEADLINE` is on the order of 2^48 (140 thousand billion)
* delete the `RandomReplacement(Spec|Benchmark)`
* perhaps also delete the (now redundant) dependency on `org.uncommons.math` (rather a separate PR, as this will not be mergable since the Inet deletion was merged)

File: akka-remote/src/main/scala/akka/remote/security/provider/AESCounterBuiltinRNG.scala
Comment (ktoso) [inline akka-remote/src/main/scala/akka/remote/security/provider/AESCounterBuiltinRNG.scala]: Please use Scala style for these, so CamelCaseForConstants.

File: akka-remote/src/test/scala/akka/remote/serialization/RandomReplacementSpec.scala
Comment (ktoso) [inline akka-remote/src/test/scala/akka/remote/serialization/RandomReplacementSpec.scala]: yeah remove at will

**Review by ktoso — APPROVED** [review]
LGTM, please apply the last changes you mentioned and let's get this in.

File: akka-remote/src/main/scala/akka/remote/security/provider/AES128CounterSecureRNG.scala
Comment (ktoso) [inline akka-remote/src/main/scala/akka/remote/security/provider/AES128CounterSecureRNG.scala]: Pool is OK, but perhaps give the thread a nice name so people know what it is? Yoiu'll need to use a ThreadFactory.

File: akka-remote/src/main/scala/akka/remote/security/provider/AESCounterBuiltinRNG.scala
Comment (ktoso) [inline akka-remote/src/main/scala/akka/remote/security/provider/AESCounterBuiltinRNG.scala]: You can mark it using the new akka.annotation.InternalApi annotation

**Comment by gosubpl** [comment]
@ktoso thanks for reviewing. I will implement all requested changes and resubmit later today.

File: akka-remote/src/main/scala/akka/remote/security/provider/AES128CounterSecureRNG.scala
Comment (gosubpl) [inline akka-remote/src/main/scala/akka/remote/security/provider/AES128CounterSecureRNG.scala]: Done

File: akka-remote/src/test/scala/akka/remote/serialization/RandomReplacementSpec.scala
Comment (gosubpl) [inline akka-remote/src/test/scala/akka/remote/serialization/RandomReplacementSpec.scala]: Tests removed. I'd rather submit a separate PR for uncommons.math dependency removal

File: akka-bench-jmh/src/main/scala/akka/remote/RandomReplacementBenchmark.scala
Comment (gosubpl) [inline akka-bench-jmh/src/main/scala/akka/remote/RandomReplacementBenchmark.scala]: Benchmark removed

File: akka-remote/src/main/scala/akka/remote/security/provider/AESCounterBuiltinRNG.scala
Comment (gosubpl) [inline akka-remote/src/main/scala/akka/remote/security/provider/AESCounterBuiltinRNG.scala]: Refactored :)

File: akka-remote/src/main/scala/akka/remote/security/provider/AESCounterBuiltinRNG.scala
Comment (gosubpl) [inline akka-remote/src/main/scala/akka/remote/security/provider/AESCounterBuiltinRNG.scala]: Added FIXME comment. I will follow with a separate PR removing `org.uncommons.math` dependency and adding `InternalApi` right after merge of this one.

**Comment by gosubpl** [comment]
Latest review remarks included. Commits squashed. IMHO ready for merge. After this is merged I will follow with a separate PR on `org.uncommons.math` removal and `InternalApi` addition as I want to avoid rebasing.
@patriknw , @ktoso many thanks for reviewing and @drewhk , @wsargent for insightful comments in previous versions of this PR!

**Comment by ktoso** [comment]
LGTM from my side, looks correct

**Comment by drewhk** [comment]
LGTM, on my side as well.

**Comment by ktoso** [comment]
Sounds like ready then, thanks a lot @gosubpl :)

**Review by patriknw — COMMENTED** [review]
Um, I left one review comment as pending yesterday. This needs a follow up PR, unless I'm wrong.

File: akka-remote/src/main/scala/akka/remote/security/provider/AESCounterBuiltinRNG.scala
Comment (gosubpl) [inline akka-remote/src/main/scala/akka/remote/security/provider/AESCounterBuiltinRNG.scala]: Thanks for spotting. I will add this in the follow-up PR.

**Pull request merged**
